keylime-ima-emulator/src/main.rs

```diff
@@ -209,25 +215,51 @@ fn main() -> std::result::Result<(), ImaEmulatorError> {
         }
     }
 
-    println!("Monitoring {}", args.ima_log.display());
+    let shutdown_marker = Arc::new(AtomicBool::new(false));
+    let marker_clone = shutdown_marker.clone();
 
-    loop {
-        for (pcr_hash_alg, position) in positions.iter_mut() {
-            *position = ml_extend(
-                &mut context,
-                &args.ima_log,
-                *position,
-                ima_hash_alg,
-                *pcr_hash_alg,
-                None,
-            )?;
+    let emulator_thread = thread::spawn(move || {
+        println!("Monitoring {}", args.ima_log.display());
+        loop {
+            // Stop if marked to shutdown
+            if marker_clone.load(Ordering::SeqCst) {
+                break;
+            }
+            for (pcr_hash_alg, position) in positions.iter_mut() {
+                *position = ml_extend(
+                    &mut context,
+                    &args.ima_log,
+                    *position,
+                    ima_hash_alg,
+                    *pcr_hash_alg,
+                    None,
+                ).expect("Error extending position {position} on PCR bank {pcr_hash_alg}");
+            }
+
+            // FIXME: We could poll IMA_ML as in the python implementation, though
+            // the file is not pollable:
+            // https://github.com/torvalds/linux/blob/master/security/integrity/ima/ima_fs.c#L267
+            // Better idea might be to check the "runtime_measurements_count" file.
```

Comment:
Maybe this can be used? https://crates.io/crates/signal-hook

If I read the code it is not using an active loop like in this solution, and the example seems really similar to the current approach.

Reply:
I can try, but I'm afraid it will fall into the case mentioned in the comment:
```
        // Do some time-limited stuff here
        // (if this could block forever, then there's no guarantee the signal will have any
        // effect).
```

Reply:
Well, apparently it works! I'll push the (much simpler) changes.

Reply:
IIRC in POSIX a singal is not a thread, is a callback that needs to be re-registered.